### DISCUSSION TITLE
How can I use mocket to test retry logic?

### DISCUSSION THREAD

**Discussion by LilyFirefly** [mindflayer/python-mocket#240] [Q&A] [answered]
I'm currently trying to convert from httpretty to mocket. Most of the conversions have been straightforward so far, but I have one unusual use case using a callback function, which I think mocket doesn't support.

```python
    from threading import Event

    import httpretty

    with httpretty.enabled():
        with open("tests/api_responses/body.json") as _body:
            body = _body.read()
        with open("tests/api_responses/headers.json") as headers:
            headers = json.load(headers)

        event = Event()
        first = True

        def callback(request, url, headers):
            nonlocal first
            if first:
                first = False
                event.wait(1)
                return 200, headers, "{}"
            return 200, headers, body

        httpretty.register_uri(
            httpretty.GET, "https://api.github.com/users/lilyfoote", body=callback
        )
```

What I want here is the first response to be delayed (by a second) so the http library attempts a retry and successfully gets the second response immediately. If there a way to achieve this?

**Comment by mindflayer** [comment]
Hi @LilyFoote,
is the client retrying a second time due to a time-out error?

**Reply by mindflayer** [reply, marked as answer]
If that's the case, Mocket gives you a way for raising errors:
https://github.com/mindflayer/python-mocket?tab=readme-ov-file#example-of-how-to-fake-socket-errors

**Reply by LilyFirefly** [reply]
Thanks! I'll give that a go!

**Reply by LilyFirefly** [reply]
This isn't working for me. I tried:

```python
    Entry.register(
        Entry.GET,
        "https://api.github.com/users/lilyfoote",
        TimeoutError(),
        Response(body=body, headers=headers, status=200),
    )
```

and

```python
    Entry.register(
        Entry.GET,
        "https://api.github.com/users/lilyfoote",
        requests.ReadTimeout(),
        Response(body=body, headers=headers, status=200),
    )
```

But both lead to the exception being caught by `requests` and re-raised as a `ConnectionError`.

The code I'm trying to mock is:

```python
def requests_retry_view(request: HttpRequest) -> HttpResponse:
    url = "https://api.github.com/users/lilyfoote"
    try:
        requests.get(url, timeout=0.2)
    except ReadTimeout:
        github_data = requests.get(url)
    return HttpResponse(github_data.content)
```

Ideally, I would get a `ReadTimeout` from the first `requests` call triggering the second call, which would then succeed.

**Reply by LilyFirefly** [reply]
Aha! I was able to get this working by calling `urllib3.exceptions.ReadTimeoutError(None, None, None)`. Thanks for the help!

**Reply by mindflayer** [reply]
The example shows:
```python
Entry.single_register(Entry.GET, url, exception=socket.error())
```
You forgot to use `exception=` and that's why Mocket is not properly raising it. Always be sure you are raising the internal exception that would make the client - `requests` - fail the way you want.

**Reply by LilyFirefly** [reply]
I don't think this applies when using the lower level `register` API, rather than `single_register`. I don't think either `register` or `Response` accept the `exception` argument.

**Reply by mindflayer** [reply]
An `exception` is used INSTEAD of a `Response`, that's why you cannot create a second wrapping the first one. I'll have a look at `register()`, but from a quick read at the code I assume it should get an `exception` the vary same way it gets a response.

**Reply by mindflayer** [reply]
I'll add a specific test later today and see how it goes.

**Reply by mindflayer** [reply]
It was like I said, it works as expected. See: https://github.com/mindflayer/python-mocket/commit/dc20d8dd59c91d90bac1a91b54b8423b9d319fa0

**Reply by LilyFirefly** [reply]
Yeah, I got it working (https://github.com/mindflayer/python-mocket/discussions/240#discussioncomment-9689518). The trick was to use the `urllib3` exception so that `requests` re-raised it as the exception my code was expecting. There wasn't any mocket specific problem at that point.

**Reply by mindflayer** [reply]
Exactly, that's why I previously mentioned:

> Always be sure you are raising the internal exception that would make the client - requests - fail the way you want.

I am glad you solved your issues. What kind of project are you using Mocket for?

**Reply by LilyFirefly** [reply]
I work on [Kolo](https://pypi.org/project/kolo/) which is a tool for analysing code as it runs. We have specific integrations for some http libraries, so we want to be able to test that we analyse those correctly.

We also have a feature that can generate integration tests for Django projects from the data we've recorded and we're switching from `httpretty` to `mocket` in those generated tests too.

**Reply by mindflayer** [reply]
What a nice product! Kudos!
